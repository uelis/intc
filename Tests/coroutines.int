/* interaction strategy for tail recursion */
fn tailrecimp(x) =
  case x of
    Inr(a, ()) -> return Inl((), Inr(a, ()))
  | Inl(u, s) ->
      case s of
         Inr(b) -> return Inr(b)
       | Inl(d, (a, ())) -> return Inl((), Inr(a, ()))

/* The tail recursion combinator is implemented by the above strategy */
let tailrec =
   direct(tailrecimp : (('a -> ['b]) -> 'a -> ['b]) -> 'a -> ['b])

/* interaction strategies for coroutines */
let coroutimpl= fn x ->
  case x of 
    Inl(ud, m) ->
        (case m of
          Inl(e, (a, ())) ->
            (case ud of 
               Inl() -> return Inr(Inl(e, Inr(a, ())))
             | Inr(d) -> return Inr(Inl(e, Inl(d, a))))
        | Inr(c) -> return Inr(Inr(c)))
  | Inr(r) -> 
      (case r of 
        Inl(e, m) ->
            (case m of
              Inl(d, (b, ())) -> 
               return Inl(Inr(d), Inl(e, b))
            | Inr(c) -> 
                return Inr(Inr(c)))
      | Inr() -> return Inl(Inl(), Inr()))

/* coroutine combinator */      
let corout = direct(coroutimpl
   : (('a -> ['b]) -> ['c]) ->
     (('b -> ['a]) -> ('a -> ['c])) ->
     ['c] )

let println = fn i -> 
   print i;
   print "\n"

/* Example */
   
let proc1 = 位 yield ->
     tailrec (位 proc1 ->
       fn j ->
         val b = intslt(500, j) in
         if b then
           return ()
         else
           print "p1: ";
           println j;
  	       val j' = intadd(j, 2) in
           val i = yield j' in
           proc1 i
       ) 3

let proc2 = 位 yield ->
   copy yield as yield1, yield2 in
   tailrec (位 proc2 ->
     fn j ->
       print "p2: ";
       println j;
       val j' = intsub(j, 1) in
       val k = yield1 j' in
       print "p2: got ";
       println k;
       val k' = intsub(k, 1) in
       val i = yield2 k' in
       proc2 i
   ) 

let main =
   corout proc1 proc2
