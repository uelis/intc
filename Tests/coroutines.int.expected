p1: 3
p2: 5
p1: 4
p2: got 6
p1: 5
p2: 7
p1: 6
p2: got 8
p1: 7
p2: 9
p1: 8
p2: got 10
p1: 9
p2: 11
p1: 10
p2: got 12
p1: 11
p2: 13
p1: 12
p2: got 14
p1: 13
p2: 15
p1: 14
p2: got 16
p1: 15
p2: 17
p1: 16
p2: got 18
p1: 17
p2: 19
p1: 18
p2: got 20
p1: 19
p2: 21
p1: 20
p2: got 22
p1: 21
p2: 23
p1: 22
p2: got 24
p1: 23
p2: 25
p1: 24
p2: got 26
p1: 25
p2: 27
p1: 26
p2: got 28
p1: 27
p2: 29
p1: 28
p2: got 30
p1: 29
p2: 31
p1: 30
p2: got 32
p1: 31
p2: 33
p1: 32
p2: got 34
p1: 33
p2: 35
p1: 34
p2: got 36
p1: 35
p2: 37
p1: 36
p2: got 38
p1: 37
p2: 39
p1: 38
p2: got 40
p1: 39
p2: 41
p1: 40
p2: got 42
p1: 41
p2: 43
p1: 42
p2: got 44
p1: 43
p2: 45
p1: 44
p2: got 46
p1: 45
p2: 47
p1: 46
p2: got 48
p1: 47
p2: 49
p1: 48
p2: got 50
p1: 49
p2: 51
p1: 50
p2: got 52
p1: 51
p2: 53
p1: 52
p2: got 54
p1: 53
p2: 55
p1: 54
p2: got 56
p1: 55
p2: 57
p1: 56
p2: got 58
p1: 57
p2: 59
p1: 58
p2: got 60
p1: 59
p2: 61
p1: 60
p2: got 62
p1: 61
p2: 63
p1: 62
p2: got 64
p1: 63
p2: 65
p1: 64
p2: got 66
p1: 65
p2: 67
p1: 66
p2: got 68
p1: 67
p2: 69
p1: 68
p2: got 70
p1: 69
p2: 71
p1: 70
p2: got 72
p1: 71
p2: 73
p1: 72
p2: got 74
p1: 73
p2: 75
p1: 74
p2: got 76
p1: 75
p2: 77
p1: 76
p2: got 78
p1: 77
p2: 79
p1: 78
p2: got 80
p1: 79
p2: 81
p1: 80
p2: got 82
p1: 81
p2: 83
p1: 82
p2: got 84
p1: 83
p2: 85
p1: 84
p2: got 86
p1: 85
p2: 87
p1: 86
p2: got 88
p1: 87
p2: 89
p1: 88
p2: got 90
p1: 89
p2: 91
p1: 90
p2: got 92
p1: 91
p2: 93
p1: 92
p2: got 94
p1: 93
p2: 95
p1: 94
p2: got 96
p1: 95
p2: 97
p1: 96
p2: got 98
p1: 97
p2: 99
p1: 98
p2: got 100
p1: 99
p2: 101
p1: 100
p2: got 102
p1: 101
p2: 103
p1: 102
p2: got 104
p1: 103
p2: 105
p1: 104
p2: got 106
p1: 105
p2: 107
p1: 106
p2: got 108
p1: 107
p2: 109
p1: 108
p2: got 110
p1: 109
p2: 111
p1: 110
p2: got 112
p1: 111
p2: 113
p1: 112
p2: got 114
p1: 113
p2: 115
p1: 114
p2: got 116
p1: 115
p2: 117
p1: 116
p2: got 118
p1: 117
p2: 119
p1: 118
p2: got 120
p1: 119
p2: 121
p1: 120
p2: got 122
p1: 121
p2: 123
p1: 122
p2: got 124
p1: 123
p2: 125
p1: 124
p2: got 126
p1: 125
p2: 127
p1: 126
p2: got 128
p1: 127
p2: 129
p1: 128
p2: got 130
p1: 129
p2: 131
p1: 130
p2: got 132
p1: 131
p2: 133
p1: 132
p2: got 134
p1: 133
p2: 135
p1: 134
p2: got 136
p1: 135
p2: 137
p1: 136
p2: got 138
p1: 137
p2: 139
p1: 138
p2: got 140
p1: 139
p2: 141
p1: 140
p2: got 142
p1: 141
p2: 143
p1: 142
p2: got 144
p1: 143
p2: 145
p1: 144
p2: got 146
p1: 145
p2: 147
p1: 146
p2: got 148
p1: 147
p2: 149
p1: 148
p2: got 150
p1: 149
p2: 151
p1: 150
p2: got 152
p1: 151
p2: 153
p1: 152
p2: got 154
p1: 153
p2: 155
p1: 154
p2: got 156
p1: 155
p2: 157
p1: 156
p2: got 158
p1: 157
p2: 159
p1: 158
p2: got 160
p1: 159
p2: 161
p1: 160
p2: got 162
p1: 161
p2: 163
p1: 162
p2: got 164
p1: 163
p2: 165
p1: 164
p2: got 166
p1: 165
p2: 167
p1: 166
p2: got 168
p1: 167
p2: 169
p1: 168
p2: got 170
p1: 169
p2: 171
p1: 170
p2: got 172
p1: 171
p2: 173
p1: 172
p2: got 174
p1: 173
p2: 175
p1: 174
p2: got 176
p1: 175
p2: 177
p1: 176
p2: got 178
p1: 177
p2: 179
p1: 178
p2: got 180
p1: 179
p2: 181
p1: 180
p2: got 182
p1: 181
p2: 183
p1: 182
p2: got 184
p1: 183
p2: 185
p1: 184
p2: got 186
p1: 185
p2: 187
p1: 186
p2: got 188
p1: 187
p2: 189
p1: 188
p2: got 190
p1: 189
p2: 191
p1: 190
p2: got 192
p1: 191
p2: 193
p1: 192
p2: got 194
p1: 193
p2: 195
p1: 194
p2: got 196
p1: 195
p2: 197
p1: 196
p2: got 198
p1: 197
p2: 199
p1: 198
p2: got 200
p1: 199
p2: 201
p1: 200
p2: got 202
p1: 201
p2: 203
p1: 202
p2: got 204
p1: 203
p2: 205
p1: 204
p2: got 206
p1: 205
p2: 207
p1: 206
p2: got 208
p1: 207
p2: 209
p1: 208
p2: got 210
p1: 209
p2: 211
p1: 210
p2: got 212
p1: 211
p2: 213
p1: 212
p2: got 214
p1: 213
p2: 215
p1: 214
p2: got 216
p1: 215
p2: 217
p1: 216
p2: got 218
p1: 217
p2: 219
p1: 218
p2: got 220
p1: 219
p2: 221
p1: 220
p2: got 222
p1: 221
p2: 223
p1: 222
p2: got 224
p1: 223
p2: 225
p1: 224
p2: got 226
p1: 225
p2: 227
p1: 226
p2: got 228
p1: 227
p2: 229
p1: 228
p2: got 230
p1: 229
p2: 231
p1: 230
p2: got 232
p1: 231
p2: 233
p1: 232
p2: got 234
p1: 233
p2: 235
p1: 234
p2: got 236
p1: 235
p2: 237
p1: 236
p2: got 238
p1: 237
p2: 239
p1: 238
p2: got 240
p1: 239
p2: 241
p1: 240
p2: got 242
p1: 241
p2: 243
p1: 242
p2: got 244
p1: 243
p2: 245
p1: 244
p2: got 246
p1: 245
p2: 247
p1: 246
p2: got 248
p1: 247
p2: 249
p1: 248
p2: got 250
p1: 249
p2: 251
p1: 250
p2: got 252
p1: 251
p2: 253
p1: 252
p2: got 254
p1: 253
p2: 255
p1: 254
p2: got 256
p1: 255
p2: 257
p1: 256
p2: got 258
p1: 257
p2: 259
p1: 258
p2: got 260
p1: 259
p2: 261
p1: 260
p2: got 262
p1: 261
p2: 263
p1: 262
p2: got 264
p1: 263
p2: 265
p1: 264
p2: got 266
p1: 265
p2: 267
p1: 266
p2: got 268
p1: 267
p2: 269
p1: 268
p2: got 270
p1: 269
p2: 271
p1: 270
p2: got 272
p1: 271
p2: 273
p1: 272
p2: got 274
p1: 273
p2: 275
p1: 274
p2: got 276
p1: 275
p2: 277
p1: 276
p2: got 278
p1: 277
p2: 279
p1: 278
p2: got 280
p1: 279
p2: 281
p1: 280
p2: got 282
p1: 281
p2: 283
p1: 282
p2: got 284
p1: 283
p2: 285
p1: 284
p2: got 286
p1: 285
p2: 287
p1: 286
p2: got 288
p1: 287
p2: 289
p1: 288
p2: got 290
p1: 289
p2: 291
p1: 290
p2: got 292
p1: 291
p2: 293
p1: 292
p2: got 294
p1: 293
p2: 295
p1: 294
p2: got 296
p1: 295
p2: 297
p1: 296
p2: got 298
p1: 297
p2: 299
p1: 298
p2: got 300
p1: 299
p2: 301
p1: 300
p2: got 302
p1: 301
p2: 303
p1: 302
p2: got 304
p1: 303
p2: 305
p1: 304
p2: got 306
p1: 305
p2: 307
p1: 306
p2: got 308
p1: 307
p2: 309
p1: 308
p2: got 310
p1: 309
p2: 311
p1: 310
p2: got 312
p1: 311
p2: 313
p1: 312
p2: got 314
p1: 313
p2: 315
p1: 314
p2: got 316
p1: 315
p2: 317
p1: 316
p2: got 318
p1: 317
p2: 319
p1: 318
p2: got 320
p1: 319
p2: 321
p1: 320
p2: got 322
p1: 321
p2: 323
p1: 322
p2: got 324
p1: 323
p2: 325
p1: 324
p2: got 326
p1: 325
p2: 327
p1: 326
p2: got 328
p1: 327
p2: 329
p1: 328
p2: got 330
p1: 329
p2: 331
p1: 330
p2: got 332
p1: 331
p2: 333
p1: 332
p2: got 334
p1: 333
p2: 335
p1: 334
p2: got 336
p1: 335
p2: 337
p1: 336
p2: got 338
p1: 337
p2: 339
p1: 338
p2: got 340
p1: 339
p2: 341
p1: 340
p2: got 342
p1: 341
p2: 343
p1: 342
p2: got 344
p1: 343
p2: 345
p1: 344
p2: got 346
p1: 345
p2: 347
p1: 346
p2: got 348
p1: 347
p2: 349
p1: 348
p2: got 350
p1: 349
p2: 351
p1: 350
p2: got 352
p1: 351
p2: 353
p1: 352
p2: got 354
p1: 353
p2: 355
p1: 354
p2: got 356
p1: 355
p2: 357
p1: 356
p2: got 358
p1: 357
p2: 359
p1: 358
p2: got 360
p1: 359
p2: 361
p1: 360
p2: got 362
p1: 361
p2: 363
p1: 362
p2: got 364
p1: 363
p2: 365
p1: 364
p2: got 366
p1: 365
p2: 367
p1: 366
p2: got 368
p1: 367
p2: 369
p1: 368
p2: got 370
p1: 369
p2: 371
p1: 370
p2: got 372
p1: 371
p2: 373
p1: 372
p2: got 374
p1: 373
p2: 375
p1: 374
p2: got 376
p1: 375
p2: 377
p1: 376
p2: got 378
p1: 377
p2: 379
p1: 378
p2: got 380
p1: 379
p2: 381
p1: 380
p2: got 382
p1: 381
p2: 383
p1: 382
p2: got 384
p1: 383
p2: 385
p1: 384
p2: got 386
p1: 385
p2: 387
p1: 386
p2: got 388
p1: 387
p2: 389
p1: 388
p2: got 390
p1: 389
p2: 391
p1: 390
p2: got 392
p1: 391
p2: 393
p1: 392
p2: got 394
p1: 393
p2: 395
p1: 394
p2: got 396
p1: 395
p2: 397
p1: 396
p2: got 398
p1: 397
p2: 399
p1: 398
p2: got 400
p1: 399
p2: 401
p1: 400
p2: got 402
p1: 401
p2: 403
p1: 402
p2: got 404
p1: 403
p2: 405
p1: 404
p2: got 406
p1: 405
p2: 407
p1: 406
p2: got 408
p1: 407
p2: 409
p1: 408
p2: got 410
p1: 409
p2: 411
p1: 410
p2: got 412
p1: 411
p2: 413
p1: 412
p2: got 414
p1: 413
p2: 415
p1: 414
p2: got 416
p1: 415
p2: 417
p1: 416
p2: got 418
p1: 417
p2: 419
p1: 418
p2: got 420
p1: 419
p2: 421
p1: 420
p2: got 422
p1: 421
p2: 423
p1: 422
p2: got 424
p1: 423
p2: 425
p1: 424
p2: got 426
p1: 425
p2: 427
p1: 426
p2: got 428
p1: 427
p2: 429
p1: 428
p2: got 430
p1: 429
p2: 431
p1: 430
p2: got 432
p1: 431
p2: 433
p1: 432
p2: got 434
p1: 433
p2: 435
p1: 434
p2: got 436
p1: 435
p2: 437
p1: 436
p2: got 438
p1: 437
p2: 439
p1: 438
p2: got 440
p1: 439
p2: 441
p1: 440
p2: got 442
p1: 441
p2: 443
p1: 442
p2: got 444
p1: 443
p2: 445
p1: 444
p2: got 446
p1: 445
p2: 447
p1: 446
p2: got 448
p1: 447
p2: 449
p1: 448
p2: got 450
p1: 449
p2: 451
p1: 450
p2: got 452
p1: 451
p2: 453
p1: 452
p2: got 454
p1: 453
p2: 455
p1: 454
p2: got 456
p1: 455
p2: 457
p1: 456
p2: got 458
p1: 457
p2: 459
p1: 458
p2: got 460
p1: 459
p2: 461
p1: 460
p2: got 462
p1: 461
p2: 463
p1: 462
p2: got 464
p1: 463
p2: 465
p1: 464
p2: got 466
p1: 465
p2: 467
p1: 466
p2: got 468
p1: 467
p2: 469
p1: 468
p2: got 470
p1: 469
p2: 471
p1: 470
p2: got 472
p1: 471
p2: 473
p1: 472
p2: got 474
p1: 473
p2: 475
p1: 474
p2: got 476
p1: 475
p2: 477
p1: 476
p2: got 478
p1: 477
p2: 479
p1: 478
p2: got 480
p1: 479
p2: 481
p1: 480
p2: got 482
p1: 481
p2: 483
p1: 482
p2: got 484
p1: 483
p2: 485
p1: 484
p2: got 486
p1: 485
p2: 487
p1: 486
p2: got 488
p1: 487
p2: 489
p1: 488
p2: got 490
p1: 489
p2: 491
p1: 490
p2: got 492
p1: 491
p2: 493
p1: 492
p2: got 494
p1: 493
p2: 495
p1: 494
p2: got 496
p1: 495
p2: 497
p1: 496
p2: got 498
p1: 497
p2: 499
p1: 498
p2: got 500
p1: 499
p2: 501
p1: 500
p2: got 502
